File: app/src/main/java/edu/umsl/bekzod_tolipov/MainActivity.java
package edu.umsl.bekzod_tolipov;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    private TextView questionTextView;
    private QuizModel model;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        questionTextView = findViewById(R.id.questionTextView);

        int index = 0;
        if(savedInstanceState != null){
            index = savedInstanceState.getInt("QUESTION_INDEX", 0);
        }

        model = new QuizModel(index);
        questionTextView.setText(model.getCurrentQuestion().getId());

        Button trueButton = findViewById(R.id.trueButton);
        Button falseButton = findViewById(R.id.falseButton);

        trueButton.setOnClickListener(userSelectionListener);
        falseButton.setOnClickListener(userSelectionListener);

    }

    View.OnClickListener userSelectionListener = new View.OnClickListener(){
        @Override
        public void onClick(View v){
            boolean selectedValue = false;
            if (v.getId() == R.id.trueButton){
                selectedValue = true;
            }

            if(model.isAnswerCorrect(selectedValue)) {
                Toast.makeText(MainActivity.this,
                        R.string.correct_toast,
                        Toast.LENGTH_SHORT)
                        .show();
            }else{
                Toast.makeText(MainActivity.this,
                        R.string.incorrect_toast,
                        Toast.LENGTH_SHORT)
                        .show();
            }
            model.advanceToNextQuestion();
            questionTextView.setText(model.getCurrentQuestion().getId());
        }
    };

    //Preferred way
//    public void clickedTrue(View view){
//        Log.d("Tag", "Clicked the true button");
//        if(model.isAnswerCorrect(true)) {
//            Toast.makeText(this, R.string.correct_toast, Toast.LENGTH_SHORT)
//                    .show();
//        }else{
//            Toast.makeText(this, R.string.incorrect_toast, Toast.LENGTH_SHORT)
//                    .show();
//        }
//    }

//    View.OnClickListener truePressed = new View.OnClickListener(){
//      @Override
//      public void onClick(View v){
//          model.isAnswerCorrect();
//      }
//    };

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt("QUESTION_INDEX", model.getCurrentIndex());
    }
}

File: app/src/main/java/edu/umsl/bekzod_tolipov/MainActivity.kt
package edu.umsl.bekzod_tolipov

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity: AppCompatActivity() {

    private lateinit var model: QuizModel
    private var isCheater: Boolean = false

    companion object{
        const val REQUEST_CODE_CHEAT = 0
        const val POSITION_KEY = "position"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        savedInstanceState?.let {bundle ->
            val index = bundle.getInt(POSITION_KEY, 0)
            model = QuizModel(index)
        }
        if (!::model.isInitialized){
            model = QuizModel(0)
        }

        trueButton.setOnClickListener(selectionListener)
        falseButton.setOnClickListener(selectionListener)
        questionTextView.setText(model.currentQuestion.id)

        cheatButton.setOnClickListener {
            val isTrue = model.currentQuestion.isTrue
            val intent = CheatActivity.newIntent(this, isTrue)
 //           startActivity(intent)
            startActivityForResult(intent, REQUEST_CODE_CHEAT)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode != Activity.RESULT_OK) {
            return
        }

        if (data == null) { return }

        when (requestCode) {
            REQUEST_CODE_CHEAT -> {
                this.isCheater = CheatActivity.wasAnswerShown(data)
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle): Unit {
        super.onSaveInstanceState(outState)
        outState.putInt("QUESTION_INDEX", model.currentIndex)
    }

    private var selectionListener = object : View.OnClickListener {
        override fun onClick(v: View) {
            var selectedValue = false
            if (v.id == R.id.trueButton) {
                selectedValue = true
            }

            val messageResId = if (isCheater) {
                R.string.judgment_toast
            } else {
                if (model.isAnswerCorrect(selectedValue)) {
                    R.string.correct_toast
                } else {
                    R.string.incorrect_toast
                }
            }

            Toast.makeText(this@MainActivity,
                    messageResId,
                    Toast.LENGTH_SHORT)
                    .show()

            isCheater = false
            model.advanceToNextQuestion()
            questionTextView.setText(model.currentQuestion.id)
        }
    }


}